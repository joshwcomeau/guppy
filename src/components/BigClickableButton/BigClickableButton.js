// @flow
import React, { Component } from 'react';
import { Spring, animated } from 'react-spring';
import styled from 'styled-components';
import Color from 'color';

import { COLORS, GRADIENTS } from '../../constants';

type Props = {
  width: number,
  height: number,
  thickness: number,
  colors: Array<string>,
  children?: React$Node,
  isOn: boolean,
  onClick: (ev: SyntheticMouseEvent<*>) => void,
};

type State = {
  isActive: boolean,
};

class BigClickableButton extends Component<Props, State> {
  static defaultProps = {
    width: 64,
    height: 64,
    thickness: 6,
    colors: GRADIENTS.darkPrimary,
  };

  state = {
    status: 'off',
    isActive: false,
  };

  depress = () => this.setState({ isActive: true });
  release = () => this.setState({ isActive: false });

  render() {
    const {
      width,
      height,
      thickness,
      colors,
      onClick,
      isOn,
      children,
    } = this.props;
    const { isActive } = this.state;

    const bottomColor = Color(colors[0])
      .darken(0.3)
      .rgb()
      .string();
    const sideColor = Color(colors[1])
      .darken(0.3)
      .rgb()
      .string();

    const background = `linear-gradient(45deg, ${colors.join(', ')})`;

    // When the user is actively clicking on it, it should be all-the-way
    // depressed (full thickness).
    // Otherwise, the amount of depression will depend on whether or not the
    // button is "on".
    // transition order is always 0 => thickness =>  thickness - 2
    const amountDepressedOptions = [0, thickness, thickness - 2];

    const depressedToIndex = isActive ? 1 : isOn ? 2 : 0;
    const depressedFromIndex =
      depressedToIndex === 0 ? 2 : depressedToIndex - 1;

    return (
      <OuterWrapper thickness={thickness}>
        <Spring
          from={{ offset: amountDepressedOptions[depressedFromIndex] }}
          to={{ offset: amountDepressedOptions[depressedToIndex] }}
          config={{
            tension: 196,
            friction: 16,
          }}
          native
        >
          {interpolated => (
            <Wrapper width={width} height={height} offset={interpolated.offset}>
              <Button
                width={width}
                height={height}
                background={background}
                onMouseDown={this.depress}
                onMouseUp={this.release}
                onMouseLeave={this.release}
                onClick={onClick}
              >
                {children}
              </Button>
              <ButtonSideWrapper>
                <SidesSvg
                  style={{ display: 'block' }}
                  width={width + thickness + 2}
                  height={height + thickness + 2}
                >
                  <path
                    d={`
                      M 0,${height}
                      L ${width},${height}
                      L ${width + thickness + 2},${height + thickness + 2}
                      L ${thickness},${height + thickness + 2}
                    `}
                    fill={bottomColor}
                  />
                  <path
                    d={`
                      M ${width},0
                      L ${width + thickness + 2},${thickness}
                      L ${width + thickness + 2},${height + thickness + 2}
                      L ${width}, ${height}
                    `}
                    fill={sideColor}
                  />
                </SidesSvg>
              </ButtonSideWrapper>
            </Wrapper>
          )}
        </Spring>
      </OuterWrapper>
    );
  }
}

const OuterWrapper = styled.div`
  position: relative;
  overflow: hidden;
  padding-right: ${props => props.thickness}px;
  padding-bottom: ${props => props.thickness}px;
`;

export const Wrapper = animated(styled.div`
  position: relative;
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  transform: ${props => `translate(${props.offset}px, ${props.offset}px)`};
`);

export const Button = styled.button`
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  display: flex;
  align-items: center;
  justify-content: space-around;
  color: ${COLORS.textOnBackground};
  font-size: 18px;

  border: none;
  background: ${props => props.background};
  outline: none; /* TODO: Better a11y story :/ */
`;

export const ButtonSideWrapper = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  pointer-events: none;
`;

const SidesSvg = styled.svg``;

export default BigClickableButton;
