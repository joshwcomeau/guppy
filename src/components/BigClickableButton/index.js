// @flow
export { default } from './BigClickableButton';
