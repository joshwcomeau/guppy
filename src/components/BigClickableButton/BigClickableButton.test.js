/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { mount } from 'enzyme';
import BigClickableButton, {
  Button,
  ButtonSideWrapper,
} from './BigClickableButton';
// TODO: Use Lolex to mock setTimeout - later use Jest mock once PR https://github.com/facebook/jest/pull/5171 landed
import lolex from 'lolex';

describe('BigClickableButton component', () => {
  let wrapper;
  let button;
  lolex.install(); // mock setTimeout

  beforeEach(() => {
    wrapper = mount(<BigClickableButton />);
    button = wrapper.find('button');
    jest.clearAllTimers();
    button.simulate('mouseDown');
  });

  const createSnapshots = () => {
    expect(wrapper.find(Button)).toMatchSnapshot();
    expect(wrapper.find(ButtonSideWrapper)).toMatchSnapshot();
  };

  it('should render pressed', () => {
    expect(wrapper.state('isActive')).toBeTruthy();
    createSnapshots();
  });

  it('should render released', () => {
    button.simulate('mouseUp');
    expect(wrapper.state('isActive')).toBeFalsy();
    createSnapshots();
  });
});
