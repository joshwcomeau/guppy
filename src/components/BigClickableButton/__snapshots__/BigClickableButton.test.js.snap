// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigClickableButton component should render pressed 1`] = `
.c0 {
  width: 64px;
  height: 64px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  color: #FFF;
  font-size: 18px;
  border: none;
  background: linear-gradient(45deg,#651fff,#304FFE);
  outline: none;
}

<styled.button
  background="linear-gradient(45deg, #651fff, #304FFE)"
  height={64}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  width={64}
>
  <button
    className="c0"
    height={64}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    width={64}
  />
</styled.button>
`;

exports[`BigClickableButton component should render pressed 2`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  pointer-events: none;
}

<styled.div>
  <div
    className="c0"
  >
    <styled.svg
      height={72}
      style={
        Object {
          "display": "block",
        }
      }
      width={72}
    >
      <svg
        className=""
        height={72}
        style={
          Object {
            "display": "block",
          }
        }
        width={72}
      >
        <path
          d="
                      M 0,64
                      L 64,64
                      L 72,72
                      L 6,72
                    "
          fill="rgb(63, 0, 200)"
        />
        <path
          d="
                      M 64,0
                      L 72,6
                      L 72,72
                      L 64, 64
                    "
          fill="rgb(1, 33, 210)"
        />
      </svg>
    </styled.svg>
  </div>
</styled.div>
`;

exports[`BigClickableButton component should render released 1`] = `
.c0 {
  width: 64px;
  height: 64px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  color: #FFF;
  font-size: 18px;
  border: none;
  background: linear-gradient(45deg,#651fff,#304FFE);
  outline: none;
}

<styled.button
  background="linear-gradient(45deg, #651fff, #304FFE)"
  height={64}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  width={64}
>
  <button
    className="c0"
    height={64}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    width={64}
  />
</styled.button>
`;

exports[`BigClickableButton component should render released 2`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  pointer-events: none;
}

<styled.div>
  <div
    className="c0"
  >
    <styled.svg
      height={72}
      style={
        Object {
          "display": "block",
        }
      }
      width={72}
    >
      <svg
        className=""
        height={72}
        style={
          Object {
            "display": "block",
          }
        }
        width={72}
      >
        <path
          d="
                      M 0,64
                      L 64,64
                      L 72,72
                      L 6,72
                    "
          fill="rgb(63, 0, 200)"
        />
        <path
          d="
                      M 64,0
                      L 72,6
                      L 72,72
                      L 64, 64
                    "
          fill="rgb(1, 33, 210)"
        />
      </svg>
    </styled.svg>
  </div>
</styled.div>
`;
